### PR TITLE
Add speech async JSON/REST system tests.

### PR DESCRIPTION
**Pull request by daspecster**
This adds JSON/REST system tests for speech's `async_recognize()` method.

### CODE REVIEW

File: system_tests/speech.py [hunk 1]
```diff
@@ -79,6 +79,21 @@ def _make_sync_request(self, content=None, source_uri=None,
                                        speech_context=['Google', 'cloud'])
         return result
 
+    def _make_async_request(self, content=None, source_uri=None,
+                            max_alternatives=None):
+        client = speech.Client(use_gax=False)
```
Comment (dhermes) [inline system_tests/speech.py]: Why `use_gax=False` shouldn't the env. running the system test be the one that decides?

File: system_tests/speech.py [hunk 2]
```diff
@@ -79,6 +79,21 @@ def _make_sync_request(self, content=None, source_uri=None,
...
+                               source_uri=source_uri,
+                               encoding=speech.Encoding.LINEAR16,
+                               sample_rate=16000)
+        result = client.async_recognize(sample,
+                                        language_code='en-US',
+                                        max_alternatives=max_alternatives,
+                                        profanity_filter=True,
+                                        speech_context=['Google', 'cloud'])
+
+        return result
```
Comment (dhermes) [inline system_tests/speech.py]: Why not just `return client.async_recognize(...)`?

File: system_tests/speech.py [hunk 3]
```diff
@@ -109,3 +124,54 @@ def test_sync_recognize_gcs_file(self):
         result = self._make_sync_request(source_uri=source_uri,
                                          max_alternatives=1)
         self._check_best_results(result)
+
+    def test_async_recognize_local_file(self):
+        import time
+        with open(AUDIO_FILE, 'rb') as file_obj:
+            operation = self._make_async_request(content=file_obj.read(),
+                                                 max_alternatives=2)
```
Comment (dhermes) [inline system_tests/speech.py]: Can you de-couple the `file_obj.read()` from the network call.

```python
with open(AUDIO_FILE, 'rb') as file_obj:
    content = file_obj.read()

operation = self._make_async_request(content=content,
                                     max_alternatives=2)
```

File: system_tests/speech.py [hunk 4]
```diff
@@ -109,3 +124,54 @@ def test_sync_recognize_gcs_file(self):
         result = self._make_sync_request(source_uri=source_uri,
                                          max_alternatives=1)
         self._check_best_results(result)
+
+    def test_async_recognize_local_file(self):
+        import time
```
Comment (dhermes) [inline system_tests/speech.py]: Import this globally?

File: system_tests/speech.py [hunk 5]
```diff
@@ -109,3 +124,54 @@ def test_sync_recognize_gcs_file(self):
...
+        with open(AUDIO_FILE, 'rb') as file_obj:
+            operation = self._make_async_request(content=file_obj.read(),
+                                                 max_alternatives=2)
+        retry_count = 100
+        while retry_count > 0 and not operation.complete:
+            retry_count -= 1
+            time.sleep(10)
+            operation.poll()
```
Comment (dhermes) [inline system_tests/speech.py]: Please use `system_tests/retry.py`

File: system_tests/speech.py [hunk 6]
```diff
@@ -109,3 +124,54 @@ def test_sync_recognize_gcs_file(self):
...
+        blob_name = 'hello.wav'
+        blob = Config.TEST_BUCKET.blob(blob_name)
+        self.to_delete_by_case.append(blob)  # Clean-up.
+        with open(AUDIO_FILE, 'rb') as file_obj:
+            blob.upload_from_file(file_obj)
+
+        source_uri = 'gs://%s/%s' % (bucket_name, blob_name)
+        operation = self._make_async_request(source_uri=source_uri,
+                                             max_alternatives=2)
+        retry_count = 100
```
Comment (dhermes) [inline system_tests/speech.py]: 100 retries! WAT?

File: system_tests/speech.py [hunk 7]
```diff
@@ -109,3 +124,54 @@ def test_sync_recognize_gcs_file(self):
...
+        while retry_count > 0 and not operation.complete:
+            retry_count -= 1
+            time.sleep(10)
```
Comment (dhermes) [inline system_tests/speech.py]: 100 * 10 = 1000 seconds?

File: system_tests/speech.py [hunk 8]
```diff
@@ -109,3 +124,54 @@ def test_sync_recognize_gcs_file(self):
...
+            operation.poll()
+
+        self.assertEqual(len(operation.results), 2)
+        self._check_best_results(operation.results)
+
+        results = operation.results
+        self.assertIsInstance(results[0], Transcript)
+        self.assertEqual(results[0].transcript, 'hello ' + self.ASSERT_TEXT)
+        self.assertGreater(results[0].confidence, 0.90)
```
Comment (dhermes) [inline system_tests/speech.py]: You already check all of lines 143, 144 and 145 in `self._check_best_results(operation.results)`

File: system_tests/speech.py (hunk 1 above)
Comment (dhermes) [inline system_tests/speech.py]: We set `passenv = GOOGLE_*` so that `GOOGLE_CLOUD_DISABLE_GRPC` gets passed through and used the way we intend.

Comment (daspecster) [inline system_tests/speech.py]: I think it already does.
```
[testenv:system-tests]
basepython =
    python2.7
commands =
    python {toxinidir}/system_tests/attempt_system_tests.py {posargs}
deps =
    {[testing]deps}
passenv = GOOGLE_* TRAVIS* encrypted_*
```

I guess my question is more how would they get run?

File: system_tests/speech.py (hunk 6 above)
Comment (daspecster) [inline system_tests/speech.py]: Yeah there doesn't need to be that many.

File: system_tests/speech.py (hunk 7 above)
Comment (daspecster) [inline system_tests/speech.py]: Using retry now so this is gone!

File: system_tests/speech.py (hunk 1 above)
Comment (daspecster) [inline system_tests/speech.py]: OK, so should I take out these tests and add that to `.travis.yml`? I guess I'm not sure what the ask is here exactly?

Comment (dhermes) [inline system_tests/speech.py]: The ask: don't use the `use_gax` param when constructing a client

File: system_tests/speech.py (hunk 5 above)
Comment (daspecster) [inline system_tests/speech.py]: Looks like I found a bug while switching.
```
Traceback (most recent call last):
  File "/Users/daspecster/Documents/google/google-cloud-py/system_tests/speech.py", line 176, in test_async_recognize_gcs_file
    _wait_until_complete(operation)
  File "/Users/daspecster/Documents/google/google-cloud-py/system_tests/speech.py", line 48, in _wait_until_complete
    return retry(operation.poll)()
  File "/Users/daspecster/Documents/google/google-cloud-py/system_tests/retry.py", line 144, in wrapped_function
    result = to_wrap(*args, **kwargs)
  File "/Users/daspecster/Documents/google/google-cloud-py/.tox/system-tests/lib/python2.7/site-packages/google/cloud/speech/operation.py", line 109, in poll
    self._update(api_response)
  File "/Users/daspecster/Documents/google/google-cloud-py/.tox/system-tests/lib/python2.7/site-packages/google/cloud/speech/operation.py", line 130, in _update
    self._metadata = Metadata.from_api_repr(metadata)
  File "/Users/daspecster/Documents/google/google-cloud-py/.tox/system-tests/lib/python2.7/site-packages/google/cloud/speech/metadata.py", line 49, in from_api_repr
    progress_percent = response['progressPercent']
KeyError: 'progressPercent'
```

File: system_tests/speech.py [hunk 9]
```diff
@@ -22,10 +22,32 @@
...
+    :type max_attempts: int
+    :param max_attempts: (Optional) The maximum number of times to check if
+                         the operation has completed. Defaults to 5.
+
+    :rtype: bool
+    :returns: Boolean indicating if the operation is complete.
+    """
+
+    def _operation_complete(result):
+        return result
```
Comment (dhermes) [inline system_tests/speech.py]: Define `_operation_complete` at module-level

File: system_tests/speech.py [hunk 10]
```diff
@@ -126,29 +147,27 @@ def test_sync_recognize_gcs_file(self):
         self._check_best_results(result)
 
     def test_async_recognize_local_file(self):
-        import time
+        if Config.USE_GAX:
+            self.skipTest('async_recognize gRPC not yet implemented.')
```
Comment (dhermes) [inline system_tests/speech.py]: Put in your tracking to remove this `skipTest` once the implementation is done?

Comment (daspecster) [inline system_tests/speech.py]: https://github.com/GoogleCloudPlatform/google-cloud-python/issues/2522

**Comment by dhermes** [comment]
LGTM (assuming you are tracking those `skipTest` usages)

File: system_tests/speech.py (hunk 9 above)
Comment (daspecster) [inline system_tests/speech.py]: https://github.com/GoogleCloudPlatform/google-cloud-python/pull/2648/commits/d45ec65e8da371c85920fa60956f9ea4933d3565

File: system_tests/speech.py [hunk 11]
```diff
@@ -22,10 +22,34 @@
...
+    :param operation: Operation that has not completed.
+
+    :type max_attempts: int
+    :param max_attempts: (Optional) The maximum number of times to check if
+                         the operation has completed. Defaults to 5.
+
+    :rtype: bool
+    :returns: Boolean indicating if the operation is complete.
+    """
+
```
Comment (dhermes) [inline system_tests/speech.py]: Hi whitespace, I'm Dan

Comment (dhermes) [inline system_tests/speech.py]: (FWIW, feel free to leave it here, but I doubt you meant to when you moved `_operation_complete` out)

Comment (daspecster) [inline system_tests/speech.py]: https://github.com/GoogleCloudPlatform/google-cloud-python/pull/2648/commits/b79b1eed8634c8a974b5ba2352d6466a6e544cab

**Comment by daspecster** [comment]
Ok I'm squashing now. Thanks!

**Pull request merged**